File: template/cpp_template/src/customized_converter.cc
// Customized converter of model, including pre-process and post-process.

#include "customized_converter.h"

#include "logger/logger.h"

namespace netease::grps {

// Should be copiable and movable.
struct YourCtxVal {
  // Add your codes here.
  std::string val;
};

YourConverter::YourConverter() = default;
YourConverter::~YourConverter() = default;

void YourConverter::Init(const std::string& path, const YAML::Node& args) {
  Converter::Init(path, args);
  // Add your codes here.
  CLOG4(INFO, "your converter init, path: " << path << ", args: " << args);
}

void YourConverter::PreProcess(const ::grps::protos::v1::GrpsMessage& input,
                               std::vector<std::pair<std::string, TensorWrapper>>& output,
                               GrpsContext& ctx) {
  // Preprocess request and convert to tensors.
  // Add your codes here.

  // Your can set context user data of current request as follows:
  YourCtxVal your_ctx_val{"your_value"};
  ctx.SetUserData(std::move(your_ctx_val)); // Use move to avoid copy.

  // Set tf tensor like:
  // std::shared_ptr<tensorflow::Tensor> tensor = std::make_shared<tensorflow::Tensor>();
  // output.emplace_back("your_tensor_name", tensor);

  // Set torch tensor like:
  // std::shared_ptr<at::Tensor> tensor = std::make_shared<at::Tensor>();
  // output.emplace_back("your_tensor_name", tensor);

  // Set eigen tensor like:
  // std::shared_ptr<Eigen::Tensor<float, 2>> tensor = std::make_shared<Eigen::Tensor<float, 2>>(1, 2);
  // (*tensor)(0, 0) = 1.0;
  // (*tensor)(0, 1) = 2.0;
  // output.emplace_back("your_tensor_name", tensor);
}

void YourConverter::PostProcess(const std::vector<std::pair<std::string, TensorWrapper>>& input,
                                ::grps::protos::v1::GrpsMessage& output,
                                GrpsContext& ctx) {
  // Postprocess tensors and convert to response.
  // Add your codes here like:

  // Your can get context user data of current request as follows:
  auto& your_value = ctx.GetUserData<YourCtxVal>(); // Use reference to avoid copy.
  CLOG4(INFO, "your context user data value: " << your_value.val);

  output.set_str_data("hello grps."); // Add string data.

  // Add generic tensor([[1.0, 2.0, 3.0]]) like:
  /*
  auto* g_tensor = output.mutable_gtensors()->add_tensors();
  // tensor name
  g_tensor->set_name("tensor_name");
  // shape [1, 3]
  g_tensor->add_shape(1);
  g_tensor->add_shape(3);
  // data type float32
  g_tensor->set_dtype(::grps::protos::v1::DT_FLOAT32);
  // data [1, 2, 3]
  g_tensor->add_flat_float32(1);
  g_tensor->add_flat_float32(2);
  g_tensor->add_flat_float32(3);
  */
}

void YourConverter::BatchPreProcess(std::vector<const ::grps::protos::v1::GrpsMessage*>& inputs,
                                    std::vector<std::pair<std::string, TensorWrapper>>& output,
                                    std::vector<GrpsContext*>& ctxs) {
  // You can preprocess every request and convert to tensor. Merge tensors of each request to batch tensor.
  // Add your codes here.
}

void YourConverter::BatchPostProcess(const std::vector<std::pair<std::string, TensorWrapper>>& input,
                                     std::vector<::grps::protos::v1::GrpsMessage*>& outputs,
                                     std::vector<GrpsContext*>& ctxs) {
  // You can postprocess batch tensor and convert to response. Split batch tensor to tensors of each request.
  // Add your codes here.
}
} // namespace netease::grps

File: template/py_template/src/customized_converter.py
# Customized converter of model, including pre-process and post-process.

import os
import sys

# Add src dir to sys.path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from grps_framework.apis.grps_pb2 import GrpsMessage, GenericTensor, DataType
from grps_framework.context.context import GrpsContext
from grps_framework.converter.converter import Converter, converter_register
from grps_framework.logger.logger import clogger


class YourConverter(Converter):
    """Your converter."""

    def init(self, path=None, args=None):
        """
        Init converter.

        Args:
            path: Path.
            args: More args.

        Raises:
            Exception: If init failed, can raise exception and exception will be caught by server and show error message
            to user when start service.
        """
        super().init(path, args)
        clogger.info('your converter init, path: {}, args: {}'.format(path, args))

    def preprocess(self, inp: GrpsMessage, context: GrpsContext):
        """
        Preprocess.

        Args:
            inp: Input message from client or previous model(multi model sequential mode).
            context: Grps context of current request.

        Returns:
            Pre-processed data which is input of model inferer.

        Raises:
            Exception: If preprocess failed, can raise exception and exception will be caught by server and return error
            message to client.
        """
        # your can set context for current request like, value can be any type:
        context.put_user_data('key', 'value')

        # preprocess request and convert to tensors.
        # add your codes here.
        return {}

    def postprocess(self, inp, context: GrpsContext) -> GrpsMessage:
        """
        Postprocess.

        Args:
            inp: Input to be post-processed, which is output of model inferer.
            context: Grps context of current request.

        Returns:
            Post-processed data with GrpsMessage format to client or next model(multi model sequential mode).

        Raises:
            Exception: If postprocess failed, can raise exception and exception will be caught by server and return error
            message to client.
        """
        # you can get context of current request like:
        clogger.info('context: {}'.format(context.get_user_data('key')))

        out = GrpsMessage()
        # postprocess tensors and convert to response.
        # set string data like:
        out.str_data = 'hello grps.'  # add string data.

        # set generic tensor([[1.0, 2.0, 3.0]] like:
        # gtensor = GenericTensor(name='tensor_name', dtype=DataType.DT_FLOAT32, shape=[1, 3], flat_float32=[1, 2, 3])
        # out.gtensors.tensors.append(gtensor)
        return out

    def batch_preprocess(self, inps: list, contexts: list):
        """
        Batch preprocess.

        Args:
            inps: Input messages from client or previous model(multi model sequential mode).
            contexts: Grps contexts of current requests.

        Returns:
            Pre-processed data which is input of model inferer.

        Raises:
            Exception: If preprocess failed, can raise exception and exception will be caught by server and return error
            message to client.
        """
        # You can preprocess every request and convert to tensor. Merge tensors of each request to batch tensor.
        # Add your codes here.
        pass

    def batch_postprocess(self, inp, contexts: list) -> list:
        """
        Batch postprocess.

        Args:
            inp: Input to be post-processed, which is output of model inferer.
            contexts: Grps contexts of current requests.

        Returns:
            Post-processed data with GrpsMessage format to client or next model(multi model sequential mode).

        Raises:
            Exception: If postprocess failed, can raise exception and exception will be caught by server and return
            error message to client.
        """
        # You can postprocess batch tensor and convert to response. Split batch tensor to tensors of each request.
        # Add your codes here.
        pass


converter_register.register('your_converter', YourConverter())
